tools/scripts/add_library_tso_doublets.py

```diff
@@ -0,0 +1,180 @@
+    print("Reading Optimus h5ad:")
+    gex_data = gex_h5ad_modified
+    # NHashID is optional input, so the logic below sets it if undefined
+    if gex_nhash_id is not None:
+        gex_data.uns['NHashID'] = gex_nhash_id
+    else:
+        gex_nhash_id = "NA"
+        gex_data.uns['NHashID'] = gex_nhash_id
+
+    #gex_data.write(f"{input_id}.h5ad")
+
+    print("Reading library metrics")
+    library = pd.read_csv(library_csv, header=None)
+
+    # Calculates total library TSO metrics
+    # TSO reads refer to reads derived from the Template Switch Oligo
+    # TSO reads per cell are calculated from the BAM cN BAM tag
+    print("Calculating TSO frac")
+    tso_reads = gex_data.obs.tso_reads.sum() / gex_data.obs.n_reads.sum()
+    print("TSO reads:")
+    print(tso_reads)
+    
+    print("Calclating keeper metrics based on doublets and n_genes")
+    if counting_mode == "sc_rna":
+        gene_threshold = 1500
+    else:
+        gene_threshold = 1000
+
+    estimated_cells = len(gex_data[gex_data.obs["star_IsCell"]==True])
+    # Expected cells is the number expected from the experiment; usually 10,000 with 10x data
+    expected_cells = int(expected_cells)  # Placeholder, replace with actual value
+    
+    # Adding doublet scores to barcodes that have been called as cells
+    all_barcodes = pd.DataFrame(index=gex_data.obs_names)
+    # Merge doublet scores with all barcodes, filling missing values with NA
+    all_barcodes = all_barcodes.join(doublet_scores, how='left')
+    # Assign the doublet scores back to the adata object
+    gex_data.obs['doublet_score'] = all_barcodes['doublet_score']
+
+    # Adding keeper metrics
+    subset = gex_data[(gex_data.obs['star_IsCell']== True) & (gex_data.obs['doublet_score']<0.3) & (gex_data.obs['n_genes']> gene_threshold)]
+    keeper_cells = subset.shape[0]
+    keeper_mean_reads_per_cell = subset.obs["n_reads"].mean()
+    keeper_median_genes = subset.obs["n_genes"].median() 
+    percent_keeper = keeper_cells/estimated_cells
+    percent_usable = keeper_cells/expected_cells
+
+    # Updating library metrics
+    dictionary = library.set_index(0)[1].to_dict()
+    dictionary['frac_tso'] = tso_reads
+    dictionary['percent_doublets'] = doublets
+    dictionary['keeper_cells'] = keeper_cells
+    dictionary['keeper_mean_reads_per_cell'] = keeper_mean_reads_per_cell
+    dictionary['keeper_median_genes'] = keeper_median_genes
+    dictionary['percent_keeper'] = percent_keeper*100
+    dictionary['percent_usable'] = percent_usable*100
+
+    new_dictionary = {"NHashID": [gex_nhash_id]}  # This line is fine, it already has a list
+    # Update other scalar values to lists
+    dictionary = {key: [value] for key, value in dictionary.items()}
+    new_dictionary.update(dictionary)
+    new_dictionary = pd.DataFrame(new_dictionary)
+    new_dictionary.transpose().to_csv("library_metrics.csv", header=None)
+    return gex_data
+
+
+if __name__ == "__main__":
```

Comment:
Consider add a `main()` function the script's logic and keep the argument parsing into the `if __name__ == "__main__":` block but move the remaining coide to the main function and call    `main(args)'

Reply:
revised to include this